src/web/releases.rs

```diff
@@ -72,67 +72,37 @@ impl Default for Order {
-                    releases.rustdoc_status,
-                    crates.github_stars
-             FROM crates
-             INNER JOIN releases ON crates.id = releases.crate_id
-             ORDER BY releases.release_time DESC
-             LIMIT $1 OFFSET $2"
-        }
-        Order::GithubStars => {
-            "SELECT crates.name,
-                    releases.version,
-                    releases.description,
-                    releases.target_name,
-                    releases.release_time,
-                    releases.rustdoc_status,
-                    crates.github_stars
-             FROM crates
-             INNER JOIN releases ON releases.id = crates.latest_version_id
-             ORDER BY crates.github_stars DESC
-             LIMIT $1 OFFSET $2"
-        }
-        Order::RecentFailures => {
-            "SELECT crates.name,
-                    releases.version,
-                    releases.description,
-                    releases.target_name,
-                    releases.release_time,
-                    releases.rustdoc_status,
-                    crates.github_stars
-             FROM crates
-             INNER JOIN releases ON crates.id = releases.crate_id
-             WHERE releases.build_status = FALSE AND releases.is_library = TRUE
-             ORDER BY releases.release_time DESC
-             LIMIT $1 OFFSET $2"
-        }
-        Order::FailuresByGithubStars => {
-            "SELECT crates.name,
-                    releases.version,
-                    releases.description,
-                    releases.target_name,
-                    releases.release_time,
-                    releases.rustdoc_status,
-                    crates.github_stars
-             FROM crates
-             INNER JOIN releases ON releases.id = crates.latest_version_id
-             WHERE releases.build_status = FALSE AND releases.is_library = TRUE
-             ORDER BY crates.github_stars DESC
-             LIMIT $1 OFFSET $2"
-        }
+    // WARNING: it is _crucial_ that this always be hard-coded and NEVER be user input
+    let (ordering, failed): (&'static str, _) = match order {
+        Order::ReleaseTime => ("releases.release_time", false),
+        Order::GithubStars => ("crates.github_stars", false),
+        Order::RecentFailures => ("releases.release_time", true),
+        Order::FailuresByGithubStars => ("crates.github_stars", true),
     };
-    let query = conn.query(&query, &[&limit, &offset]).unwrap();
+    let where_clause = if failed {
+        "WHERE releases.build_status = FALSE AND releases.is_library = TRUE"
+    } else {
+        ""
+    };
+
+    let query = format!(
+        "SELECT crates.name,
+            releases.version,
+            releases.description,
+            releases.target_name,
+            releases.release_time,
+            releases.rustdoc_status,
+            crates.github_stars
+        FROM crates
+        INNER JOIN releases ON crates.id = releases.crate_id
+        {}
```

Comment:
This could be something like

```sql
WHERE $3 OR (releases.build_status = FALSE AND releases.is_library = TRUE)
```

and pass `!failed` as a parameter.

Reply:
Good idea, updated.